### ISSUE TITLE
Multiple issues on Bash with `declare -n`

### ISSUE THREAD

**Issue by S4mu3LJ3f** [koalaman/shellcheck#3380]
- My shellcheck version (`0.11.0` and "online"):
- [x] The rule's wiki page does not already cover this
- [x] I tried on https://www.shellcheck.net/

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
declare i=0
declare -A arr_123=()
declare -n arr=arr_123

arr+=([${i}]="$i")
echo "${arr[${i}]}"
```

#### Here's what shellcheck currently says:
```
In test.sh line 4:
declare -A arr_123=()
           ^-----^ SC2034 (warning): arr_123 appears unused. Verify use (or export if used externally).

In test.sh line 7:
arr+=([${i}]="$i")
       ^--^ SC2004 (style): $/${} is unnecessary on arithmetic variables.
```
And when I change to just `[i]` behavior changes too.

Also I have issue with `SC2178` on `declare -n` that I can't reproduce in test script for now.
```sh
In scr line 895:
    declare -n array_inscr="array_inscr_${i}"   # Array was declared with `-a`
               ^---------^ SC2178 (warning): Variable was used as an array but is now assigned a string.
```

#### Here's what I wanted or expected to see:
No warnings when used without a reference.
```sh
#!/bin/bash
declare i=0
declare -A arr_123=()
#declare -n arr=arr_123

arr_123+=([${i}]="$i")
echo "${arr_123[${i}]}"
```

**Comment by wileyhy** [comment]
Hi,
  One bug per issue, please, would really help with tracking any varying
resolutions.
  Dynamic parameter names can be quite fragile, however. Multiple indexed
arrays with synchronized indices can be used for simulating
multi-dimensional arrays, depending on your use case.
  Happy Holidays,

**Comment by ghost** [comment]
I assume this is the one bug with `declare -n`.

**Comment by giner** [comment]
This issue seems to be the same as https://github.com/koalaman/shellcheck/issues/817 and being reported regularly.
